mitodl/open-discussions
# Location for profiles

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #1399

#### What's this PR do?
- Adds an input field on the profile edit form for entering a city.
- Uses the Algolia Places API to look up and display possible matches for the user's input
- Saves the entire JSON object for the selected match to the `Profile` model
- Displays the 'value' attribute of the JSON object on the profile page.

#### How should this be manually tested?
- [Sign up for an Algolia account](https://www.algolia.com/users/sign_up) and create/get your app id and a search-only api key
- Set your `ALGOLIA_APP_ID` and `ALGOLIA_API_KEY` .env values
- Go to your profile page.  It should look the same as usual.
- Edit the profile.  There should be a new field with "Location (city)" as help text.
- Start typing in a city name.  You should soon see a list of suggestions.
- Choose a suggested city and submit the form.
- The city you chose should appear on the profile page underneath your profile image.
- Edit the profile again.  Clear out the location field and submit.
- The profile page should no longer show a location.
- Edit the profile again.  Enter a nonexistent city name with no match and submit.
- The profile page should display your fake city.

#### Any background context you want to provide?
The location information returned from the API is being saved as a single JSON field in the `Profile` model for now, because it is still uncertain how much of the location data we will need and how we want to use it in the future.  Once that becomes clearer, it might make more sense to have a separate `Location` model with defined fields.

@pdpinch @Ferdi the `Built by Algolia` blurb under the suggestions list is mandatory according to the terms of service if we use the `places.js` library.  Also, the matched location text is not always consistent about the formatting it uses.  Most of the time it displays `<city>, <state/province>, <country>` but there are some notable exceptions such as New York City, which just displays `New York, United States of America`.

#### Screenshots (if appropriate)

<img width="558" alt="screen shot 2018-12-10 at 1 05 42 pm" src="https://user-images.githubusercontent.com/187676/49809576-4652fe80-fd2d-11e8-9454-a50ea82972b0.png">

<img width="388" alt="screen shot 2018-12-10 at 1 05 50 pm" src="https://user-images.githubusercontent.com/187676/49809595-5074fd00-fd2d-11e8-82b6-494f8be1ef28.png">

<img width="751" alt="screen shot 2018-12-10 at 12 45 52 pm" src="https://user-images.githubusercontent.com/187676/49809601-54088400-fd2d-11e8-91e1-33fe8d91a395.png">

<img width="376" alt="screen shot 2018-12-10 at 12 46 26 pm" src="https://user-images.githubusercontent.com/187676/49809605-5834a180-fd2d-11e8-8429-87d2f27228b8.png">
